### PR TITLE
Feature/storage/share lease

### PR DESCRIPTION
**Pull request by seanmcc-msft**

### CODE REVIEW

File: sdk/storage/Azure.Storage.Files.Shares/src/ShareLeaseClient.cs [hunk 1]
```diff
@@ -79,15 +96,43 @@ public ShareLeaseClient(ShareFileClient client, string leaseId = null)
...
+        /// <summary>
+        /// Ensure either the File or Share is present.
+        /// </summary>
+        private void EnsureClient()
+        {
+            if (FileClient == null && ShareClient == null)
+            {
+                // This can only happen if someone's not being careful while mocking
+                throw ShareErrors.FileOrShareMissing(nameof(ShareLeaseClient), nameof(FileClient), nameof(ShareClient));
+            }
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Files.Shares/src/ShareLeaseClient.cs:130]: Should we validate that only one client is present ?

On the other hand shouldn't we extract common pieces somewhere and have ShareLeaseClient and FileLeaseClient?
 - it seems we branch every piece of internal logic depending on whether it's file or share (so volume of code won't change significantly if we do this.
 - some parameters are not functional for file, so we could make it stricter that way.

**Review by kasobol-msft — COMMENTED** [review]
overall looking good.

File: sdk/storage/Azure.Storage.Files.Shares/src/ShareClient.cs [hunk 2]
```diff
@@ -749,11 +810,15 @@ private void SetNameFieldsIfNull()
         /// A <see cref="RequestFailedException"/> will be thrown if
         /// a failure occurs.
         /// </remarks>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+#pragma warning disable AZC0002 // DO ensure all service methods, both asynchronous and synchronous, take an optional CancellationToken parameter called cancellationToken.
         public virtual Response<bool> DeleteIfExists(
-            bool includeSnapshots = true,
-            CancellationToken cancellationToken = default) =>
+#pragma warning restore AZC0002 // DO ensure all service methods, both asynchronous and synchronous, take an optional CancellationToken parameter called cancellationToken.
+            bool includeSnapshots,
```
Comment (gapra-msft) [inline sdk/storage/Azure.Storage.Files.Shares/src/ShareClient.cs:817]: Should we expose this as bool? What if they add more Delete Options?

File: sdk/storage/Azure.Storage.Files.Shares/src/ShareLeaseClient.cs (hunk 1 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.Shares/src/ShareLeaseClient.cs:130]: This is pattern matched from [BlobLeaseClient](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/storage/Azure.Storage.Blobs/src/BlobLeaseClient.cs), I think it is important that we keep the SDKS internally consistent.

File: sdk/storage/Azure.Storage.Files.Shares/src/ShareClient.cs (hunk 2 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Files.Shares/src/ShareClient.cs:817]: I think it's fine for now, if they add another optional parameter, I'll switch to `ShareDeleteOptions`.

**Review by gapra-msft — APPROVED** [review]

**Review by kasobol-msft — APPROVED** [review]

**Pull request merged**